src/coprocessor/endpoint.rs

```diff
@@ -416,12 +415,15 @@ impl<E: Engine> Endpoint<E> {
 
         tracker.on_begin_all_items();
 
-        let handle_request_future = track(handler.handle_request(), &mut tracker);
-        let result = if let Some(semaphore) = &semaphore {
+        let deadline = tracker.req_ctx.deadline;
+        let handle_request_future =
+            check_deadline(track(handler.handle_request(), &mut tracker), deadline);
+        let deadline_res = if let Some(semaphore) = &semaphore {
             limit_concurrency(handle_request_future, semaphore, LIGHT_TASK_THRESHOLD).await
         } else {
             handle_request_future.await
         };
+        let result = deadline_res.map_err(Error::from).and_then(|res| res);
```

Comment:
Why `and_then`?

Reply:
If `deadline_res` is `Ok`, deadline is not exceeded, the result should just be the inner result.
If `deadline_res` is `Err`, the inner result is dropped and the result is the deadline error.